Simple clang fixes

delcypher opened the pull request:
This PR fixes a large proportion of the benchmarks so they can be compiled with Clang when using the ``Makefile`` build system. The majority of the commits just fix the signature of the entry point (``void main()`` -> ``int main()``).

There are two directories that still don't build ``ldv-linux-3.4-simple`` and ``regression`` with Clang 3.7. All other directories build.

I am going to merge this as soon as I see that compilation passes under GCC.

PhilippWendler commented:
Is it really necessary to change all this? Does Clang not support a void main function? As I see it, the C standard allows main to have return type void.

delcypher commented:
> Is it really necessary to change all this? Does Clang not support a void main function? As I see it, the C standard allows main to have return type void

http://stackoverflow.com/questions/2108192/what-are-the-valid-signatures-for-cs-main-function

Clang does not think it is valid C99. It will accept it if the standard is set to ``gnu99`` rather than ``c99`` so I'm inclined to think that ``void main(void)`` is a GCC extension and therefore should be discouraged.

The build system is set to use ``c99`` rather than ``gnu99`` unless it is absolutely necessary.

PhilippWendler commented:
The C standard says that a non-int main functions are allowed as implementation-defined behavior, so this makes sense.

Note that the SV-Comp rules [explicitly state that benchmarks are GNUC](http://sv-comp.sosy-lab.org/2016/rules.php#programs), I would not expect to be able to compile all with a strict C99 setting.

delcypher commented:
> Note that the SV-Comp rules explicitly state that benchmarks are GNUC, I would not expect to be able to compile all with a strict C99 setting.

Yes I saw that but I would argue we should build strict C99 when possible, especially now that we have two production quality C compilers at our disposal rather than one.

Unless you (or anyone else) have a strong reason to prevent me from making the benchmarks more compliant with C99 then I will merge this.

dbeyer commented:
> Unless you (or anyone else) have a strong reason to prevent me from making the benchmarks more compliant with C99 then I will merge this.

Let's please merge via pull request and let's wait until tomorrow with the actual merge.
I personally have no strong opinion on this: I am with Philipp that we should not make unnecessary changes,
especially if it's just because clang cannot deal with it, on the other hand the changes do not hurt much I think.

BTW: There are not only two production quality C parsers: there is also (perhaps a bit outdated though) CIL, and there is the Eclipse CDT, which is also not too bad.
All can parse GNU C, only clang seems to suck here, as there are so many people complaining about issues with clang.

Best,
Dirk

delcypher commented:
> Let's please merge via pull request and let's wait until tomorrow with the actual merge.
Sure I can wait but is there any reason to do so? The build for gcc passes for this PR.

> I personally have no strong opinion on this: I am with Philipp that we should not make unnecessary changes,
especially if it's just because clang cannot deal with it, on the other hand the changes do not hurt much I think.

Clang **can deal** with GNU C, it's just that the build system deliberately tries to be C99 compliant not "GNU C99" compliant most of the time.

> BTW: There are not only two production quality C parsers: there is also (perhaps a bit outdated though) CIL, and there is the Eclipse CDT, which is also not too bad.
All can parse GNU C, only clang seems to suck here, as there are so many people complaining about issues with clang.

I actually meant C Compiler in the sense of a full compiler (source -> binary), not just a C parser. In any case it is not Clang that sucks here. It's the other way round.

* Clang is being more compliant to the C99 standard than GCC. Both compilers are being passed but ``-std=c99`` but only Clang is flagging up the ``void main()`` issue. Clang will accept ``void main()`` if passed ``-std=gnu99``

* The reason so many people are complaining with issues relating with Clang is **not because Clang sucks**. It's because **the benchmarks contain bad code**. The authors of the benchmarks most likely didn't try using Clang and consequently didn't see the (frequently superior, although with gcc5 they are becoming comparable) warnings/errors that Clang reports. As with other cases where Clang has been introduced (http://clang.debian.net/ http://llvm.linuxfoundation.org/) changes need to be made to sources to make them more compliant with the programming language standard they were written in so that Clang will compile them.

* Clang has an in-process assembler and thus when passed ``-fsyntax-only`` will actually find problems with assembly syntax whereas gcc will not.

* "GNU C99" is not a standard, C99 is. Tying a project to a single compiler and its (might not be documented) specific non standard behaviour is a bad idea. Therefore using C99 rather than GNUC99 when possible is desirable.

dbeyer commented:
I meant waiting to see if anybody else has objections.

On October 29, 2015 11:28:40 PM GMT+01:00, Dan Liew <notifications@github.com> wrote:
>> Let's please merge via pull request and let's wait until tomorrow
>with the actual merge.
>Sure I can wait but is there any reason to do so? The build for gcc
>passes for this PR.
>
>> I personally have no strong opinion on this: I am with Philipp that
>we should not make unnecessary changes,
>especially if it's just because clang cannot deal with it, on the other
>hand the changes do not hurt much I think.
>
>Clang **can deal** with GNU C, it's just that the build system
>deliberately tries to be C99 compliant not "GNU C99" compliant most of
>the time.
>
>> BTW: There are not only two production quality C parsers: there is
>also (perhaps a bit outdated though) CIL, and there is the Eclipse CDT,
>which is also not too bad.
>All can parse GNU C, only clang seems to suck here, as there are so
>many people complaining about issues with clang.
>
>I actually meant C Compiler in the sense of a full compiler (source ->
>binary), not just a C parser. In any case it is not Clang that sucks
>here. It's the other way round.
>
>* Clang is being more compliant to the C99 standard than GCC. Both
>compilers are being passed but ``-std=c99`` but only Clang is flagging
>up the ``void main()`` issue. Clang will accept ``void main()`` if
>passed ``-std=gnu99``
>
>* The reason so many people are complaining with issues relating with
>Clang is **not because Clang sucks**. It's because **the benchmarks
>contain bad code**. The authors of the benchmarks most likely didn't
>try using Clang and consequently didn't see the (frequently superior,
>although with gcc5 they are becoming comparable) warnings/errors that
>Clang reports. As with other cases where Clang has been introduced
>(http://clang.debian.net/ http://llvm.linuxfoundation.org/) changes
>need to be made to sources to make them more compliant with the
>programming language standard they were written in so that Clang will
>compile them.
>
>* Clang has an in-process assembler and thus when passed
>``-fsyntax-only`` will actually find problems with assembly syntax
>whereas gcc will not.
>
>* "GNU C99" is not a standard, C99 is. Tying a project to a single
>compiler and its (might not be documented) specific non standard
>behaviour is a bad idea. Therefore using C99 rather than GNUC99 when
>possible is desirable.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/dbeyer/sv-benchmarks/pull/87#issuecomment-152345753

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

PhilippWendler commented:
> "GNU C99" is not a standard, C99 is. Tying a project to a single compiler and its (might not be documented) specific non standard behaviour is a bad idea. Therefore using C99 rather than GNUC99 when possible is desirable.

This is a nice goal, but it is not our position to enforce it. We want to have benchmark programs from real projects, and these projects often use GNU C code (e.g., the Linux kernel uses GNU extensions heavily).
I prefer having benchmarks with GNU C code instead of not having these benchmarks at all.

We can of course fix trivial cases like here, but there may well be other GNU-specific stuff we can never get around, so the question is whether this is worth it compared to simply using `-std=gnu99`.

delcypher commented:
> I prefer having benchmarks with GNU C code instead of not having these benchmarks at all.

I agree with that sentiment which is why the build system allows the C standard to be overridden on a directory by directory basis. If you run ``git grep CC.Standard`` you will see where the default is set (``Makefile.config``) and which directories this has be overriden in.

> We can of course fix trivial cases like here, but there may well be other GNU-specific stuff we can never get around, so the question is whether this is worth it compared to simply using -std=gnu99.

We could change the default to ``gnuc99`` in the future but right now using ``c99`` as much as possible will make the benchmarks more compliant with C99. I think this is something the community needs to decide.

delcypher commented:
> I meant waiting to see if anybody else has objections.

I see. Well I've not seen any objections yet.

In the meantime, I've added an additional commit that enables Clang for TravisCI too. It doesn't build all the benchmarks right now so I added a hack for TravisCI that lets it skip the two directories that don't build correctly right now.

Provided the build passes with this change I will ahead and merge.

delcypher commented:
@dbeyer Okay the build using Clang (+ the ``CLANG_BUILD_HACK``) has passed. Could you please merge this?
